File: lib/api/forwardTest.ts
export type ForwardTestRun = {
  id: number;
  symbol: string;
  start_time: string;
  end_time: string | null;
  is_active: boolean;
  summary: string | null;
};

export type ForwardTestTrade = {
  id: number;
  direction: "long" | "short";
  entry_price: number;
  stop_loss: number;
  take_profit: number;
  exit_price?: number | null;
  exit_reason?: string | null;
  r_multiple?: number | null;
  profit_loss?: number | null;
  drawdown?: number | null;
  candle_time: string;
};

export type ForwardTestStatus = {
  run: ForwardTestRun;
  open_trades: ForwardTestTrade[];
  recent_trades: ForwardTestTrade[];
};

export type ForwardTestTradesResponse = {
  trades: ForwardTestTrade[];
};

function getBackendBaseUrl(): string {
  const url = process.env.NEXT_PUBLIC_BACKEND_URL;
  if (!url) throw new Error("NEXT_PUBLIC_BACKEND_URL not set");
  return url.replace(/\/+$/, "");
}

export async function startForwardTest(symbol: string): Promise<ForwardTestRun> {
  const base = getBackendBaseUrl();
  const res = await fetch(`${base}/api/v1/forward-test/start?symbol=${encodeURIComponent(symbol)}`, {
    method: "POST",
  });
  if (!res.ok) {
    const text = await res.text();
    throw new Error(`Start forward test failed: ${res.status} ${text}`);
  }
  return res.json();
}

export async function fetchForwardTestStatus(testRunId: number): Promise<ForwardTestStatus> {
  const base = getBackendBaseUrl();
  const res = await fetch(`${base}/api/v1/forward-test/status?test_run_id=${testRunId}`, { cache: "no-store" });
  if (!res.ok) {
    const text = await res.text();
    throw new Error(`Forward test status failed: ${res.status} ${text}`);
  }
  return res.json();
}

export async function fetchForwardTestTrades(testRunId: number, signal?: AbortSignal): Promise<ForwardTestTradesResponse> {
  const base = getBackendBaseUrl();
  const res = await fetch(`${base}/api/v1/forward-test/trades?test_run_id=${testRunId}`, {
    cache: "no-store",
    signal,
  });
  if (!res.ok) {
    const text = await res.text();
    throw new Error(`Forward test trades failed: ${res.status} ${text}`);
  }
  return res.json();
}

export async function downloadForwardTestCsv(testRunId: number): Promise<Blob> {
  const base = getBackendBaseUrl();
  const res = await fetch(`${base}/api/v1/forward-test/export?test_run_id=${testRunId}`);
  if (!res.ok) {
    const text = await res.text();
    throw new Error(`Forward test export failed: ${res.status} ${text}`);
  }
  return res.blob();
}

File: backend/app/api/v1/endpoints/forward_test.py
from fastapi import APIRouter, Depends, HTTPException, Query
from fastapi.responses import StreamingResponse
from sqlalchemy.orm import Session
from datetime import datetime, timezone
import csv
import io
from typing import Optional

from ...db import get_db
from ...models.forward_test import ForwardTestRun, ForwardTestTrade
from ...services.forward_test import start_test_run

router = APIRouter()


@router.post("/start", summary="Start a 5-day forward test run")
def start_forward_test(symbol: str, db: Session = Depends(get_db)):
	run = start_test_run(db, symbol=symbol)
	return {
		"id": run.id,
		"symbol": run.symbol,
		"start_time": run.start_time,
		"end_time": run.end_time,
		"is_active": run.is_active,
	}


@router.get("/status", summary="Get forward test status")
def get_forward_test_status(
	test_run_id: int = Query(..., description="ID of the forward test run"),
	db: Session = Depends(get_db),
):
	run: Optional[ForwardTestRun] = db.query(ForwardTestRun).filter(ForwardTestRun.id == test_run_id).first()
	if not run:
		raise HTTPException(status_code=404, detail="Forward test run not found")

	open_trades = db.query(ForwardTestTrade).filter(
		ForwardTestTrade.test_run_id == run.id,
		ForwardTestTrade.exit_price.is_(None),
	).order_by(ForwardTestTrade.id.desc()).all()

	closed_trades = db.query(ForwardTestTrade).filter(
		ForwardTestTrade.test_run_id == run.id,
		ForwardTestTrade.exit_price.is_not(None),
	).order_by(ForwardTestTrade.id.desc()).limit(20).all()

	return {
		"run": {
			"id": run.id,
			"symbol": run.symbol,
			"start_time": run.start_time,
			"end_time": run.end_time,
			"is_active": run.is_active,
			"summary": run.summary_json,
		},
		"open_trades": [
			{
				"id": t.id,
				"direction": t.direction,
				"entry_price": t.entry_price,
				"stop_loss": t.stop_loss,
				"take_profit": t.take_profit,
				"candle_time": t.candle_time,
			}
			for t in open_trades
		],
		"recent_trades": [
			{
				"id": t.id,
				"direction": t.direction,
				"entry_price": t.entry_price,
				"stop_loss": t.stop_loss,
				"take_profit": t.take_profit,
				"exit_price": t.exit_price,
				"exit_reason": t.exit_reason,
				"r_multiple": t.r_multiple,
				"profit_loss": t.profit_loss,
				"drawdown": t.drawdown,
				"candle_time": t.candle_time,
			}
			for t in closed_trades
		],
	}


@router.get("/trades", summary="Get all trades for a forward test run")
def get_forward_test_trades(
	test_run_id: int = Query(..., description="ID of the forward test run"),
	db: Session = Depends(get_db),
):
	run: Optional[ForwardTestRun] = db.query(ForwardTestRun).filter(ForwardTestRun.id == test_run_id).first()
	if not run:
		raise HTTPException(status_code=404, detail="Forward test run not found")

	trades = db.query(ForwardTestTrade).filter(
		ForwardTestTrade.test_run_id == run.id
	).order_by(ForwardTestTrade.id.asc()).all()

	return {
		"trades": [
			{
				"id": t.id,
				"direction": t.direction,
				"entry_price": t.entry_price,
				"stop_loss": t.stop_loss,
				"take_profit": t.take_profit,
				"exit_price": t.exit_price,
				"exit_reason": t.exit_reason,
				"r_multiple": t.r_multiple,
				"profit_loss": t.profit_loss,
				"drawdown": t.drawdown,
				"candle_time": t.candle_time,
			}
			for t in trades
		]
	}


@router.get("/export", summary="Export forward test trades as CSV")
def export_forward_test(
	test_run_id: int = Query(..., description="ID of the forward test run"),
	db: Session = Depends(get_db),
):
	run: Optional[ForwardTestRun] = db.query(ForwardTestRun).filter(ForwardTestRun.id == test_run_id).first()
	if not run:
		raise HTTPException(status_code=404, detail="Forward test run not found")

	trades = db.query(ForwardTestTrade).filter(
		ForwardTestTrade.test_run_id == run.id
	).order_by(ForwardTestTrade.id.asc()).all()

	output = io.StringIO()
	writer = csv.writer(output)
	writer.writerow([
		"id",
		"timestamp",
		"symbol",
		"direction",
		"entry_price",
		"stop_loss",
		"take_profit",
		"exit_price",
		"exit_reason",
		"r_multiple",
		"profit_loss",
		"drawdown",
		"candle_time",
		"test_run_id",
	])
	for t in trades:
		writer.writerow([
			t.id,
			t.created_at.isoformat() if t.created_at else "",
			t.symbol,
			t.direction,
			t.entry_price,
			t.stop_loss,
			t.take_profit,
			t.exit_price if t.exit_price is not None else "",
			t.exit_reason or "",
			t.r_multiple if t.r_multiple is not None else "",
			t.profit_loss if t.profit_loss is not None else "",
			t.drawdown if t.drawdown is not None else "",
			t.candle_time.isoformat() if t.candle_time else "",
			t.test_run_id,
		])

	output.seek(0)
	filename = f"forward_test_{run.id}_{datetime.now(tz=timezone.utc).date()}.csv"
	return StreamingResponse(
		iter([output.getvalue().encode("utf-8")]),
		media_type="text/csv",
		headers={"Content-Disposition": f'attachment; filename="{filename}"'},
	)


